### ISSUE TITLE
`.nyc_output/out.json` is an empty object when using with `@cypress/react`

### ISSUE THREAD

**Issue by chrisrzhou** [cypress-io/code-coverage#472]
**Logs and screenshots**
Please provide debug logs by running Cypress from the terminal with `DEBUG=code-coverage` environment variable set, see the [Debugging](https://github.com/cypress-io/code-coverage#debugging) section of the README file.

<details>
<summary>Debug logs</summary>

```
  code-coverage reset code coverage in interactive mode +0ms
  code-coverage parsed sent coverage +32ms
  code-coverage wrote coverage file /Users/chrisrzhou/Github/cypress-coverage-repo/.nyc_output/out.json +1ms
  code-coverage parsed sent coverage +25ms
  code-coverage wrote coverage file /Users/chrisrzhou/Github/cypress-coverage-repo/.nyc_output/out.json +0ms
⚠️ file /Users/chrisrzhou/Github/cypress-coverage-repo/.nyc_output/out.json has no coverage information
Did you forget to instrument your web application? Read https://github.com/cypress-io/code-coverage#instrument-your-application
  code-coverage ⚠️ file /Users/chrisrzhou/Github/cypress-coverage-repo/.nyc_output/out.json has no coverage information +8s
  code-coverage ⚠️ file /Users/chrisrzhou/Github/cypress-coverage-repo/.nyc_output/out.json has no coverage information +1ms
  code-coverage calling NYC reporter with options { 'report-dir': '/Users/chrisrzhou/Github/cypress-coverage-repo/coverage', reporter: [ 'lcov', 'clover', 'json', 'json-summary' ], extension: [ '.js', '.cjs', '.mjs', '.ts', '.tsx', '.jsx' ], excludeAfterRemap: false, 'temp-dir': '/Users/chrisrzhou/Github/cypress-coverage-repo/.nyc_output', tempDir: '/Users/chrisrzhou/Github/cypress-coverage-repo/.nyc_output', reportDir: '/Users/chrisrzhou/Github/cypress-coverage-repo/coverage' } +7ms
  code-coverage current working directory is /Users/chrisrzhou/Github/cypress-coverage-repo +0ms
  code-coverage after reporting, returning the report folder name /Users/chrisrzhou/Github/cypress-coverage-repo/coverage +40ms
  code-coverage Final coverage in /Users/chrisrzhou/Github/cypress-coverage-repo/coverage/coverage-final.json +0ms
  code-coverage There are 0 key(s) in /Users/chrisrzhou/Github/cypress-coverage-repo/coverage/coverage-final.json +0ms
```
</details>

**Versions**

- What is this plugin's version? **@cypress/code-coverage@3.9.7 (latest)**
- What is Cypress version? **cypress@7.6.0**, **@cypress/react@5.9.1**, **@cypress/webpack-dev-server@1.4.0** (see `package.json` in repo for more details).
- What is your operating system? **Mac**
- What is the shell? **zsh**
- What is the Node version? **v12.22.1**
- What is the NPM version? **6.14.12**
- How do you instrument your application? **Through the official docs' recommendation.  See repo below for details.**
- When running tests, if you open the web application in regular browser, and open DevTools, do you see `window.__coverage__` object? **Following the official docs, I can see `window.__coverage__` object for the files that the test is touching.**
![image](https://user-images.githubusercontent.com/4425362/123366558-41da4400-d52d-11eb-9526-f1137d332d71.png)
- Is there `.nyc_output` folder? Is there `.nyc_output/out.json` file. Is it empty? **Both folder and file exists, the contents is just the empty object `{}`**.
- Do you have any custom NYC settings in `package.json` (`nyc` object) or in other [NYC config files](https://github.com/istanbuljs/nyc#configuration-files) **No**
- Do you run Cypress tests in a Docker container? **No, just locally**

**Describe the bug**
With the repo linked below, and following the official docs for component testing and code coverage, I am unable to collect coverage for the tests that are run.  The tests run fine, but the coverage `out.json` is always an empty object.  I can see the code-coverage steps in the interactive test runner.

**Link to the repo**

**Comment by chrisrzhou** [comment]
Related issues that I've tried out for the last few days but without success:
- https://github.com/cypress-io/code-coverage/issues/461

To give context on the actual project where I first encountered the bug:
- I'm building a component library (does not actually use React).
- The code is authored as an ESM module (hence the repro repo being explicit about it in `package.json` and also having to use `.cjs` for some Cypress related setup to get the test runner working).
- I'm using `@cypress/react` and the component test runner for its delightful testing experience.  That has been superb, it's just the last stretch of figuring out coverage that I'm hitting a wall.  As a result, while React is not a dependency of the library I'm writing, it is a dev dependency for my tests.  I definitely do not use `create-react-app`, and therefore have a custom `.webpack.config.cjs` and `.babelrc` in the `cypress/` folder.

I don't fully understand why code coverage is not working, but some suspicions I have if they are helpful with the debugger:
- Maybe it's related to my package being ESM-only? (i.e. `"type": "module" required in `package.json`).
- Maybe `@cypress/react` does not interoperate well with `@cypress/code-coverage` and the earlier point?
- Maybe it's related to peer dependency versions that I have installed (I simply installed the latest).
- Maybe I'm completely missing something despite spending days iterating on variations from the official docs (which has a very simple setup!).

Thank you guys for the great work, and let me know if you need any other information if the repro repo is insufficient.

**Comment by twhoff** [comment]
Hi Chris, saw your comment.

I had a look at your code base and wonder if you might be missing `babel-plugin-istanbul` from your webpack config (the one you are loading in the plugin `index.js` file?

Just a hunch - let me know how you go.

**Comment by chrisrzhou** [comment]
Hi @twhoff appreciate you taking the time to help here! 🤗

I have `istanbul` defined in `.babelrc` in https://github.com/chrisrzhou/cypress-code-coverage-issue-472/tree/main/cypress, and also have `babel-plugin-istanbul` installed as a dev dependency in `package.json`.  I have to define my own `webpack.config.cjs` since this is not a CRA (it's a component library).  The `*.cjs` cypress files are needed because the source code is written as ESM (i.e. `type: "module"` in `package.json`) and without specifying this, Cypress will complain.

**Comment by namespace-github** [comment]
Hello @chrisrzhou, hello all,

here is my **solution**, which from my point of view is a gap in the documentation.

In my case when using the new component tests, which was mentioned in the context of the question (`open-ct` or `run-ct`) `window.__coverage__` was not evaluated from `@cypress/code-coverage`.

The reason was that, according to the documentation, `import '@cypress/code-coverage/support';` only included in `support/index.js`. Cypress, however, uses in my environement instead `/support/unit.js` for component tests. Both `support/index.js` and `support/unit.js` files imports the `support/commands.js`. 

I put the line there, then **it works** as well, because cypress can register the tasks in component tests.

**Comment by chrisrzhou** [comment]
Hey @namespace-github, thank you so much for getting back to me.  I tried your solution and it does not seem to work.

In my repro repo, I added in https://github.com/chrisrzhou/cypress-code-coverage-issue-472/commit/f6e496d105774c48047f229407ebb0d8bbe0ecc2, a commit that adds `./support/unit.js` with the import.  I reran `npm test` and still observed `.nyc_output/out.json` to be an empty object.

If you have any insights or can reproduce by cloning the repro repo, that would be sweet!
https://github.com/chrisrzhou/cypress-code-coverage-issue-472

**Comment by VaibhavP17** [comment]
Is there any update on this issue ? I am facing the same issue. While running the application using dev server on regular browser, I see `window.__coverage__` object but while running the cypress test cases, I get the errors: **`cypress\.nyc_output\out.json has no coverage information`** and `"Could not find any coverage information in your application by looking at the window coverage object. Did you forget to instrument your application? See [code-coverage#instrument-your-application](https://github.com/cypress-io/code-coverage#instrument-your-application) [@cypress/code-coverage]"`

**Comment by kubijo** [comment]
Same here as @VaibhavP17 said…

**Comment by DuudeXX8** [comment]
Same...

**Comment by winjay-yu-awx** [comment]
Same here

**Comment by aautcq** [comment]
I have been struggling with an empty object in my `.nyc_output/out.json`, and I'd like to share the root cause in my code, it case it might be of use for someone else.

My code was correctly instrumented by istambul, and the Cypress code-coverage plugin was correctly set up as per the docs. I had an exclusion glob pattern in my Cypress config file, like this:

```ts
export default defineConfig({
  e2e: {
    specPattern: [
      '**/*.feature',
      '!**/ignored/**/*.feature',
    ],
  },
})
```

It turns out that

1. Even if this does not break Cypress, I should have been following the Cypress doc and use the `excludeSpecPattern` option instead to exclude test files
2. `@cypress/code-coverage` actually uses the `specPattern` option to ignore test files when computing the coverage, and the `'!**/ignored/**/*.feature'` pattern resulted in the plugin ignoring all my source files 🙃 

**TL;DR**
Don't use negative patterns in `specPattern`, prefer `excludeSpecPattern`
